### ISSUE TITLE
Move all torch.Tensor methods to codegen

### ISSUE THREAD

**Issue by hameerabbasi** [pytorch/pytorch#40215]
## 🚀 Feature
The proposal is to move all (or as many as possible) of `torch.Tensor` methods as possible into codegen.

cc: @ezyang who proposed this and @kurtamohler and @Baranowski who expressed interest.

Checklist:

#### Normal methods/properties

- [ ] `register_hook`
- [ ] `retain_grad`
- [ ] `is_shared`
- [ ] `share_memory_`
- [ ] `norm`
- [ ] `lu`
- [ ] `istft`
- [ ] `resize`
- [ ] `resize_as`
- [ ] `split`
- [ ] `unique`
- [ ] `unique_consecutive`
- [ ] `refine_names`
- [ ] `align_to`
- [ ] `unflatten`
- [ ] `rename_`
- [ ] `rename`
- [ ] `_update_names`
- [ ] `grad`

#### Dunder methods
(Determine if possible/feasible)

- [ ] `__deepcopy__`
- [ ] `__reduce_ex__`
- [ ] `__setstate__`
- [ ] `__repr__`
- [ ] `__repr__`
- [ ] `__format__`
- [ ] `__len__`
- [ ] `__hash__`
- [ ] `__dir__`
- [ ] `__array_wrap__`
- [ ] `__cuda_array_interface__`
- [ ] Operators, e.g. `__add__`.

### `tools/autograd/templates/python_variable_methods.cpp`
(In my call with @ezyang it was mentioned that C++ methods were probably hard to port, I found this to be not necessarily true. See the section "Pitch" below.)

- [ ] `THPVariable__is_view`
- [ ] `THPVariable_apply_`
- [ ] `THPVariable_stride`
- [ ] `THPVariable_get_device`
- [ ] `THPVariable_has_names`
- [ ] `THPVariable_data_ptr`
- [ ] `THPVariable_storage_offset`
- [ ] `THPVariable_dim`
- [ ] `THPVariable_numel`
- [ ] `THPVariable_contiguous`
- [ ] `THPVariable_copy_`
- [ ] `THPVariable_*_scalar`
- [ ] `THPVariable_invert`
- [ ] `THPVariable_cpu`
- [ ] `THPVariable_nonzero`
- [ ] `THPVariable_cuda`
- [ ] `THPVariable_{datatype}`
- [ ] `THPVariable_element_size`
- [ ] `THPVariable_numpy`
- [ ] `THPVariable_requires_grad_`
- [ ] `THPVariable_is_contiguous`
- [ ] `THPVariable_map_`
- [ ] `THPVariable_map2_`
- [ ] `THPVariable_new_*`
- [ ] `THPVariable_storage`
- [ ] `THPVariable_storage_type`
- [ ] `THPVariable_to`
- [ ] `THPVariable_type`
- [ ] `__len__`
- [ ] `__hash__`
- [ ] `__dir__`
- [ ] `__array_wrap__`
- [ ] `__cuda_array_interface__`

## Motivation
Resulting from a verbal discussion with @ezyang regarding #37091, it was proposed that we move all `torch.Tensor` methods and, if possible, properties to codegen. This was motivated by the large diffs in the files `torch/tensor.py` and `tools/autograd/templates/python_variable_methods.cpp` in that same PR. If all of these methods are moved to codegen, future changes could be more easily handled, as well as reducing the diff in that PR.

## Pitch
In my mind, all of these methods can be divided into (broadly) three categories:

1. Those that can be moved.
2. Those that require changes to codegen to move a large amount (e.g. if properties are added to codegen, it'll move a large amount of properties already)
3. Those that cannot be moved without significant effort, e.g. each requiring their own changes to codegen.

My proposal is we determine which methods/properties fall under which category, and then tackle moving things in categories 1 and 2.

**Comment by kurtamohler** [comment]
Hey @hameerabbasi , apologies if you mentioned this in our meetings and I missed it, but I was wondering if you have a plan how exactly to move `torch/tensor.py` into codegen. I'm happy to start helping.

Since many of the methods in that file do more than simply call another method, it seems that we'd want a scheme that's flexible enough to allow each method to continue to have specialized logic, but also enable the ability to automatically insert the same (or very similar) code into each function (like the code that your PR #37091 inserts). We could potentially use a template version of `tensor.py` to accomplish that.

For instance, right now, this method is in `torch/tensor.py`:
```
    def norm(self, p="fro", dim=None, keepdim=False, dtype=None):
        r"""See :func:`torch.norm`"""
        return torch.norm(self, p, dim, keepdim, dtype=dtype)
```

If we had a template `tools/templates/tensor.py` instead, we could use some template expression to create the change you want to make in #37091, kind of like this:
```
    def norm(self, p="fro", dim=None, keepdim=False, dtype=None):
        r"""See :func:`torch.norm`"""
        ${torch_function_handler(norm, self, [p, dim, keepdim, dtype])}
        return torch.norm(self, p, dim, keepdim, dtype=dtype)
```

Then we could use a codegen script to automatically fill in that `${...}` expression with:
```
        relevant_args = (self,)
        from torch._overrides import has_torch_function, handle_torch_function
        if type(self) is not Tensor and has_torch_function(relevant_args):
            return handle_torch_function(Tensor.norm, relevant_args, self, p=p, dim=dim, keepdim=keepdim, dtype=dtype)
```
Does something like this make sense, or do you have something else in mind?

**Comment by ezyang** [comment]
An implicit part of the goal here is to bind these functions directly from C++, using similar machinery to how the other functions are bound. This would be more performant, and reduce the number of "ways" we have to handle various methods.

**Comment by kurtamohler** [comment]
Ah, I see. That makes more sense, thanks.

**Comment by kurtamohler** [comment]
It seems like it would be fairly straightforward to move some of the dunder operators like `__rdiv__` and `__rsub__` out of `python_variable_methods.cpp` into codegen. I can start working on those.

Is the aim also to attempt to remove entries from the `PyMethodDef variable_methods[]` list in `tools/autograd/templates/python_variable_methods.cpp`, and make those be codegened as well? Or does that not matter? For instance, the entries for `__add__` and `__radd__` could potentially be generated by the `gen_python_functions.py` script, but I'm not sure if there's an advantage for doing that, since it's just linking those dunder methods to the already autogened `THPVariable_add` function.

**Comment by ezyang** [comment]
Let me talk to @rgommers about this tomorrow and then I will get back to you on this.

However, the general idea behind this task is to reduce the amount of special casing, so that e.g., if you add `__torch_function__` support, there are less places you have to edit to make things work

**Comment by rgommers** [comment]
The take-away of that discussion was: having these methods in C++ would in general be nice, but it's not a blocker for gh-37091 to land, and doesn't really change the most important aspects of that design. I'll write a more extensive summary on gh-37091.
